### ISSUE TITLE
[spi_device] Access fault when reading from command, address, and payload FIFOs 

### ISSUE THREAD

**Issue by alphan** [lowRISC/opentitan#11782]
I tried configuring `CMD_INFO_11` register for `SECTOR_ERASE` (opcode: 0x20, 3-byte address, upload: true, busy: true), sent the command over SPI and printed the fields of the `UPLOAD_STATUS` register. cmdfifo and addrfifo fields looked OK but when I try to read from the command FIFO I get an access fault:

```
I00135 ottf_isrs.c:89] Load/Store fault, mtval holds the fault address.
I00136 ottf_isrs.c:90] MTVAL value is 0x40051e00
```

I see the same behavior for address and payload FIFOs using a pretty recent bitstream (`2c1f6029d`).

cc @a-will @cfrantz

**Comment by eunchan** [comment]
Thanks I will check if anything wrong in the TL-UL converting logic inside SPI_DEVICE.

**Comment by eunchan** [comment]
The design should be OK. If HW is accessing the DPSRAM through SYS clock port, the logic lowers the SW access and hold it until HW releases the request (by getting the response from DPSRAM). Basically, the access pattern does not change across the entire DPSRAM (0x4005_1000 to 0x4005_1FFF). 

Could you please check IOPMP setting? I only can think that may be the reason.

**Comment by eunchan** [comment]
Ah.. Could you please initialize the Payload / CmdFIFO/ Addr FIFO ? I think parity error may occur if SW accesses the region.

**Comment by alphan** [comment]
Thanks @eunchan, I was able to read from the FIFOs using UPLOAD_CMDFIFO and UPLOAD_ADDRFIFO registers,. I'm not sure if the data coming out is correct for more complex cases but I tried sending SECTOR_ERASE (0x20) and PAGE_PROGRAM (0x02) back to back and the second read from UPLOAD_CMDFIFO sometimes returned 0x0 and sometimes 0x20. Might worth a closer look.

Re. the payload buffer and access fault issue: I spent some time debugging this because it looks like the payload buffer is not reset when the chip is reset. Tried this a couple of times reprogramming the fpga and I get an access fault only if I read past the payload depth (or try to read it when it's empty). We cannot use word reads by rounding up the payload depth but it works if I handle the tail correctly with byte reads. PLMK if you plan to relax this.

From what I observed, the payload buffer always has the last payload. Can you confirm? If that's the case we can't really handle the multiple commands with payloads in the command FIFO. I don't think it's something we want to (host should poll the BUSY bit) but just wanted to point it out.

**Comment by eunchan** [comment]
> 
> Thanks @eunchan <https://github.com/eunchan>, I was able to read from the FIFOs using UPLOAD_CMDFIFO and UPLOAD_ADDRFIFO registers,. I'm not sure if the data coming out is correct for more complex cases but I tried sending SECTOR_ERASE (0x20) and PAGE_PROGRAM (0x02) back to back and the second read from UPLOAD_CMDFIFO sometimes returned 0x0 and sometimes 0x20. Might worth a closer look.
> 
> 
Thanks. I think there's some timing issue. Returning 0x0 may indicate the FIFO is empty, which is weird as the FIFO not empty interrupt is generated after the HW fills the FIFO first. Worth to take a look. Thanks for finding the issue.
> Re. the payload buffer and access fault issue: I spent some time debugging this because it looks like the payload buffer is not reset when the chip is reset. Tried this a couple of times reprogramming the fpga and I get an access fault only if I read past the payload depth (or try to read it when it's empty). We cannot use word reads by rounding up the payload depth but it works if I handle the tail correctly with byte reads. PLMK if you plan to relax this.
> 
As the fault occurs due to the integrity feature in the DPSRAM, the upload function can't do anything about this. HW must update only a byte at a time. It does not know when the command is finished, so as soon as it receives the 8th beat, it writes to the DPSRAM to not lose the data. So, it cannot update the entire word with correct parity bit. So, only thing FW can do, in this case, is to initialize the entire payload portion (256B) at the beginning.
> From what I observed, the payload buffer always has the last payload. Can you confirm? If that's the case we can't really handle the multiple commands with payloads in the command FIFO. I don't think it's something we want to (host should poll the BUSY bit) but just wanted to point it out.
> 
Yes. Only one payload command can be stored. And based on the discussion with @a-will, I think SW wants to block the upload function entirely when it receives a upload command, which sets BUSY bit. SO, maybe we may not need multiple command/address slots in the FIFO.

**Comment by alphan** [comment]
>> Re. the payload buffer and access fault issue: I spent some time debugging this because it looks like the payload buffer is not reset when the chip is reset. Tried this a couple of times reprogramming the fpga and I get an access fault only if I read past the payload depth (or try to read it when it's empty). We cannot use word reads by rounding up the payload depth but it works if I handle the tail correctly with byte reads. PLMK if you plan to relax this.

> As the fault occurs due to the integrity feature in the DPSRAM, the upload function can't do anything about this. HW must update only a byte at a time. It does not know when the command is finished, so as soon as it receives the 8th beat, it writes to the DPSRAM to not lose the data. So, it cannot update the entire word with correct parity bit. So, only thing FW can do, in this case, is to initialize the entire payload portion (256B) at the beginning.

Yes, initializing the payload area solves this issue but is there a particular reason why the spi_device buffer is not reset when the chip is reset? Shouldn't the retention SRAM be the only memory whose contents survive a reset?

**Comment by eunchan** [comment]
SRAM macro usually does not have reset signals. That's why the SRAM portion needs to be initialized at the beginning. While reset, if power keeps being provided, SRAM contents remain. If we need to guarantee to have clean state of SRAM after every reset, HW should have self-initialization logic inside, which I think some what needed eventually.

**Comment by eunchan** [comment]
@alphan please check and close if no further question you have.

**Comment by alphan** [comment]
Thanks @eunchan. @cfrantz, it looks like spi_device SRAM will not be reset when the chip is reset.

**Comment by cfrantz** [comment]
We should memset it to zeros if the hardware doesn't do that for us.
